### ISSUE TITLE
feasibility of assessing 8d joint distribution

### ISSUE THREAD

**Issue by mobeets** [mobeets/nullSpaceControl#182]
read those papers larry suggested
look into implementation
etc.

**Comment by mobeets** [comment]
Kolmogorov-Smirnov two-sample test compares whether two empirical distributions are equal by taking the maximum different between their empirical cdfs. E.g., below, red and blue below are two example empirical distributions:

![screen shot 2016-05-26 at 2 16 56 pm](https://cloud.githubusercontent.com/assets/1677179/15585275/894f44ea-234c-11e6-8bcd-8df1d6330f6f.png)

but there's no natural way of extending this to multiple dimensions.

[This](http://www3.stat.sinica.edu.tw/statistica/oldpdf/a17n412.pdf) Peter Hall paper (rec'd by Larry) and [this](http://stats.stackexchange.com/questions/25946/goodness-of-fit-for-2d-histograms) StackExchange question both seem to suggest fitting a kernel density estimator (KDE) to each distribution and then comparing them by taking the L2 norm of their differences. (Eqn 2.9 of the Hall paper gives a Monte Carlo way of doing hypothesis testing that the distributions are the same.) Also Peter Hall mentions how you want to use the same smoothing parameter for both KDEs.

**Comment by mobeets** [comment]
code for (an old version of?) mmd: http://people.kyb.tuebingen.mpg.de/arthur/mmd.htm

[MMD](http://www.jmlr.org/papers/volume13/gretton12a/gretton12a.pdf) (Smola et al) is a generalization of the L2 metric used in the KDE approach to an RKHS space. it is also shown to be a generalization of the Kolmogorov-Smirnov test and "earth-mover distances."

"Our test statistic is the maximum mean discrepancy (MMD), defined as the maximum deviation in the expectation of a function evaluated on each of the random variables, taken over a sufficiently rich function class: in our case, a reproducing kernel Hilbert space (RKHS). Equivalently, the statistic can be written as the norm of the difference between distribution feature means in the RKHS. We do not require density estimates as an intermediate step."

MMD picks a sigma bandwidth value like this:
"Pick sigma such that the exponent of `exp(- ||x-y|| / (2*sigma2))`,  in other words `||x-y|| / (2*sigma2)`,  equals 1 for the median distance x and y of all distances between points from both data sets X and Y."

**Comment by mobeets** [comment]
also, with regard to KDEs over histograms: "The smoothness of the kernel density estimate is evident compared to the discreteness of the histogram, as kernel density estimates converge faster to the true underlying density for continuous random variables.[6]" ([src](https://en.wikipedia.org/wiki/Kernel_density_estimation#Definition))

![screen shot 2016-05-26 at 2 53 52 pm](https://cloud.githubusercontent.com/assets/1677179/15586309/b831b2f2-2351-11e6-965e-6271aef73423.png)

There's also an "adaptive" KDE approach ([src](https://en.wikipedia.org/wiki/Variable_kernel_density_estimation)) that's supposed to be good for high dimensions.

**Comment by mobeets** [comment]
No multivariate KDE methods exist for Matlab. Here's a really great [post](https://jakevdp.github.io/blog/2013/12/01/kernel-density-estimation/) on comparing methods available in Python. Scikit-learn's might be the best in terms of speed given sample size.

One key question, however, is whether or not to use the same bandwidth for every KDE across hypotheses? According to Peter Hall this is ideal, but not sure how to do this yet. MMD above gave a heuristic.

**Comment by mobeets** [comment]
in the MMD paper, section 3.3.1 introduces the L2 distance approach from Anderson et al (1994).

The Hall paper explains why you want to use the same smoothing parameter in the KDEs: " Such a choice ensures that, under the null hypothesis that the two samples of curves come from identical populations, the main effects of differing observation-time distributions and differing subsample sizes (for the different curves) cancel. As a result, the effect that smoothing has on bias is an order of magnitude less than it would be if different bandwidths, tuned to the respective curves, were used"

**Comment by mobeets** [comment]
According to the python post, it seems that scikit-learn's implementation is best. Here's how to select the bandwidth for some 1d data `x`:

``` python
from sklearn.grid_search import GridSearchCV
grid = GridSearchCV(KernelDensity(),
                    {'bandwidth': np.linspace(0.1, 1.0, 30)},
                    cv=20) # 20-fold cross-validation
grid.fit(x[:, None])
print grid.best_params_
```

And here's how to display it (for 1d at least) on the grid `x_grid`:

``` python
kde = grid.best_estimator_
pdf = np.exp(kde.score_samples(x_grid[:, None]))

fig, ax = plt.subplots()
ax.plot(x_grid, pdf, linewidth=3, alpha=0.5, label='bw=%.2f' % kde.bandwidth)
ax.hist(x, 30, fc='gray', histtype='stepfilled', alpha=0.3, normed=True)
ax.legend(loc='upper left')
ax.set_xlim(-4.5, 3.5);
```

![screen shot 2016-05-26 at 4 16 03 pm](https://cloud.githubusercontent.com/assets/1677179/15588715/2dcdf0c4-235d-11e6-888c-ec746f21e5fc.png)

**Comment by mobeets** [comment]
Common bandwidth choice from Hall:

"One approach to common bandwidth choice is to use a “favourite” method to compute an empirical bandwidth for each curve Xi and Yj , and then take the average value to be the common bandwidth. Another technique, appropriate in the case of plug-in rules, is to use an average value of each of the components of a plug-in bandwidth selector, and assemble the average values, using the plug-in formula, to form the common bandwidth. A third approach, valid in the context of cross-validation, is to use a criterion which is the average of the cross-validatory criteria corresponding to the different curves. For each of these methods, “average” might be defined in a weighted sense, where the weights represent the respective subsample sizes."

**Comment by mobeets** [comment]
So here's the run-down:
- Given a true Y and estimates Yh1, Yh2, ...
- Find the absolute min and max of dimensions across all Ys
- Make an x_grid spanning those mins and maxes in all dims
- Call scikit-learn and fit KDE for Y. return evaluation on x_grid
- Repeat calls to sciki-learn to fit KDEs to all Yh1, Yh2, ... using the bandwidth from Y. return evaluation on x_grid
- calculate error metric on x_grid for each Yh1, Yh2, ...

**Comment by mobeets** [comment]
write in Python, bootstrap random samples from grid to evaluate

**Comment by mobeets** [comment]
execution time to calculate kdeError for two hypotheses compared to true kde increases as size of latents goes up. for comparison, 20120601 has 18,000 timepoints. that means this will take _at least_ 5 mins, and that's just for two hyps.

![screen shot 2016-05-27 at 12 01 13 pm](https://cloud.githubusercontent.com/assets/1677179/15613689/bda08036-2402-11e6-8a2d-5bf8579614a2.png)

**Comment by mobeets** [comment]
two ideas to make this shorter:
- sample 2000 timepoints or so to fit kdes, and repeat 10 times and report avg score across reps
- only fit the true latent kde, and repeatedly evaluate resampled _hypotheses_

**Comment by mobeets** [comment]
okay, well, doing the 2nd option of the above still takes a while, but it does appear to give significantly different scores:

i split the true data in half, and bootstrapped from the testing half five times. that's bar 1. bar 2 is pruning fit 5 times.

![screen shot 2016-05-27 at 3 06 40 pm](https://cloud.githubusercontent.com/assets/1677179/15618562/b2a2023a-241c-11e6-9519-3bf3955ce58c.png)

**Comment by mobeets** [comment]
and, actually the same is true if you bootstrap instead the timepoints (used both to fit the kde and to take from the hypotheses). this one only needed 5 reps of 1000 samples

![screen shot 2016-05-27 at 3 19 00 pm](https://cloud.githubusercontent.com/assets/1677179/15618850/57de6972-241e-11e6-86b0-8be833fd46d7.png)

**Comment by mobeets** [comment]
![screen shot 2016-05-27 at 7 26 48 pm](https://cloud.githubusercontent.com/assets/1677179/15623402/f54a97f4-2440-11e6-8ccb-5da476bdc986.png)

**Comment by mobeets** [comment]
i'm gonna conclude this is infeasible.

i have it implemented, but it's going to take a lot of work to know how to interpret the various scoring systems i've come up with, and it doesn't seem very straightforward, so i might suggest we table this option and stick with mean/variance.
